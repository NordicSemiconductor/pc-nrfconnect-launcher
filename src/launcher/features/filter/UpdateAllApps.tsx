/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import Button from 'react-bootstrap/Button';

import { isWithdrawn } from '../../../ipc/apps';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { updateDownloadableApp } from '../apps/appsEffects';
import { getUpdatableVisibleApps } from '../apps/appsSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const updatableApps = useLauncherSelector(getUpdatableVisibleApps);

    const updateAllApps = () =>
        updatableApps.forEach(app => {
            if (!isWithdrawn(app)) {
                dispatch(updateDownloadableApp(app, app.latestVersion));
            }
        });

    if (updatableApps.length === 0) return null;

    return (
        <Button
            variant="outline-secondary"
            onClick={updateAllApps}
            className="me_32px"
        >
            Update all apps
        </Button>
    );
};
