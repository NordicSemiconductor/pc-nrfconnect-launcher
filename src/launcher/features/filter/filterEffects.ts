/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { Settings } from '../../../ipc/settings';
import type { AppDispatch } from '../..';
import {
    setAllShownSources,
    setNameFilter,
    setShownStates,
} from './filterSlice';

export const initializeFilters =
    (settings: Settings) => (dispatch: AppDispatch) => {
        dispatch(setShownStates(settings.appFilter.shownStates));
        dispatch(setNameFilter(settings.appFilter.nameFilter));
        dispatch(setAllShownSources(settings.appFilter.shownSources));
    };
