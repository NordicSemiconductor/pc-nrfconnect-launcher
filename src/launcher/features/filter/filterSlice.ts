/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { App } from '../../../ipc/apps';
import { SourceName } from '../../../ipc/sources';
import type { RootState } from '../..';

type ShownStates = {
    available: boolean;
    installed: boolean;
};

export type State = {
    shownSources: Set<SourceName>;
    nameFilter: string;
    shownStates: ShownStates;
};

const initialState: State = {
    shownSources: new Set(),
    nameFilter: '',
    shownStates: {
        available: true,
        installed: true,
    },
};

const slice = createSlice({
    name: 'filter',
    initialState,
    reducers: {
        setAllShownSources(
            state,
            { payload: shownSources }: PayloadAction<SourceName[]>
        ) {
            state.shownSources = new Set(shownSources);
        },
        showSource(state, { payload: sourceName }: PayloadAction<SourceName>) {
            state.shownSources.add(sourceName);
        },
        hideSource(state, { payload: sourceName }: PayloadAction<SourceName>) {
            state.shownSources.delete(sourceName);
        },
        setNameFilter(state, { payload: nameFilter }: PayloadAction<string>) {
            state.nameFilter = nameFilter;
        },
        setShownStates(
            state,
            { payload: shownStates }: PayloadAction<Partial<ShownStates>>
        ) {
            state.shownStates = {
                ...state.shownStates,
                ...shownStates,
            };
        },
    },
});

export default slice.reducer;

export const {
    setAllShownSources,
    showSource,
    hideSource,
    setNameFilter,
    setShownStates,
} = slice.actions;

export const getShownSources = (state: RootState) => state.filter.shownSources;
export const getNameFilter = (state: RootState) => state.filter.nameFilter;
export const getShownStates = (state: RootState) => state.filter.shownStates;

const matchesSourceFilter = (app: App, state: RootState) =>
    state.filter.shownSources.has(app.source);

const matchesNameFilter = (app: App, state: RootState) => {
    const filter = state.filter.nameFilter;

    try {
        return new RegExp(filter, 'i').test(app.displayName);
    } catch (_) {
        // Ignore faulty regexps
    }
    return app.displayName?.includes(filter);
};

const matchesStateFilter = (app: App, state: RootState) =>
    (state.filter.shownStates.installed && app.isInstalled) ||
    (state.filter.shownStates.available && !app.isInstalled);

export const getAppsFilter = (state: RootState) => (app: App) =>
    matchesSourceFilter(app, state) &&
    matchesNameFilter(app, state) &&
    matchesStateFilter(app, state);
