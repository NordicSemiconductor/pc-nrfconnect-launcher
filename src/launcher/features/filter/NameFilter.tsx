/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React, { useRef } from 'react';
import Form from 'react-bootstrap/Form';
import { useHotKey } from 'pc-nrfconnect-shared';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { getNameFilter, setNameFilter } from './filterSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const filter = useLauncherSelector(getNameFilter);
    const searchFieldRef = useRef<HTMLInputElement>(null);

    useHotKey({
        hotKey: ['mod+e', 'mod+f', 'mod+l'],
        title: 'Focus search field',
        isGlobal: false,
        action: () => {
            searchFieldRef.current?.focus();
            searchFieldRef.current?.select();
        },
    });

    return (
        <Form.Control
            type="text"
            placeholder="Search..."
            value={filter}
            ref={searchFieldRef}
            onChange={event => {
                const nameFilter = event.target.value;
                dispatch(setNameFilter(nameFilter));
            }}
            onClick={() => searchFieldRef.current?.select()}
        />
    );
};
