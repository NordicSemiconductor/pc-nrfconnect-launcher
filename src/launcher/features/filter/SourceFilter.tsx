/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';

import {
    hideSource as hideSourceInSettings,
    showSource as showSourceInSettings,
} from '../../../ipc/settings';
import { getAllSourceNamesSorted } from '../../reducers/appsReducer';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { getShownSources, hideSource, showSource } from './filterSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const allSourceNames = useLauncherSelector(getAllSourceNamesSorted);
    const shownSourceNames = useLauncherSelector(getShownSources);

    return (
        <Col className="pl-4 pr-0">
            <div className="border-bottom py-1 mx-3 mb-2">Sources</div>
            {allSourceNames.map((sourceName, i) => (
                <Form.Check
                    label={sourceName}
                    id={`cb-${sourceName}`}
                    key={`cb-${i + 1}`}
                    className="mx-3 py-1 px-4 text-capitalize"
                    custom
                    checked={shownSourceNames.has(sourceName)}
                    onChange={({ target }) => {
                        if (target.checked) {
                            dispatch(showSource(sourceName));
                            showSourceInSettings(sourceName);
                        } else {
                            dispatch(hideSource(sourceName));
                            hideSourceInSettings(sourceName);
                        }
                    }}
                />
            ))}
        </Col>
    );
};
