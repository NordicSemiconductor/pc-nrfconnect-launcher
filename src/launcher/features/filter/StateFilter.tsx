/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';

import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import { getShownStates, setShownStates } from './filterSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const { installed, downloadable } = useLauncherSelector(getShownStates);

    return (
        <Col className="pl-0 pr-4">
            <div className="border-bottom mx-3 mb-2 py-1">State</div>
            <Form.Check
                label="Installed"
                id="cb-installed"
                className="mx-3 px-4 py-1"
                custom
                checked={installed}
                onChange={({ target }) => {
                    dispatch(
                        setShownStates({
                            installed: target.checked,
                            downloadable,
                        }),
                    );
                }}
            />
            <Form.Check
                label="Downloadable"
                id="cb-downloadable"
                className="mx-3 px-4 py-1"
                custom
                checked={downloadable}
                onChange={({ target }) => {
                    dispatch(
                        setShownStates({
                            installed,
                            downloadable: target.checked,
                        }),
                    );
                }}
            />
        </Col>
    );
};
