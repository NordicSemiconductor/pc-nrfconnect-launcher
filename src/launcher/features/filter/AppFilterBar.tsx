/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React, { useState } from 'react';
import Dropdown from 'react-bootstrap/Dropdown';
import Row from 'react-bootstrap/Row';

import NameFilter from './NameFilter';
import SourceFilter from './SourceFilter';
import StateFilter from './StateFilter';
import UpdateAllApps from './UpdateAllApps';

const FilterDropdown = () => {
    const [active, toggleActive] = useState(false);

    return (
        <Dropdown onToggle={() => toggleActive(!active)}>
            <Dropdown.Toggle variant="outline-secondary" active={active}>
                <span className="mdi mdi-tune" />
                Filter
            </Dropdown.Toggle>
            <Dropdown.Menu>
                <Row className="flex-nowrap">
                    <SourceFilter />
                    <StateFilter />
                </Row>
            </Dropdown.Menu>
        </Dropdown>
    );
};

export default () => (
    <div className="filterbox w-100 d-inline-flex">
        <FilterDropdown />
        <NameFilter />
        <div className="flex-fill" />
        <UpdateAllApps />
    </div>
);
