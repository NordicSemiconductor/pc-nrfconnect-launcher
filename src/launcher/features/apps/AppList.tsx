/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import { App as AppType, isInstalled } from '../../../ipc/apps';
import { useLauncherSelector } from '../../util/hooks';
import WithScrollbarContainer from '../../util/WithScrollbarContainer';
import AppFilterBar from '../filter/AppFilterBar';
import { getAppsFilter } from '../filter/filterSlice';
import ReleaseNotesDialog from '../releaseNotes/ReleaseNotesDialog';
import App from './App/App';
import { getAllApps } from './appsSlice';
import ConfirmLaunchDialog from './ConfirmLaunchDialog';
import InstallOtherVersionDialog from './InstallOtherVersionDialog';

const sortByStateAndName = (appA: AppType, appB: AppType) => {
    const cmpInstalled = Number(isInstalled(appB)) - Number(isInstalled(appA));

    const aName = appA.displayName || appA.name;
    const bName = appB.displayName || appB.name;

    return cmpInstalled || aName.localeCompare(bName);
};

export default () => {
    const allApps = useLauncherSelector(getAllApps);
    const appsFilter = useLauncherSelector(getAppsFilter);

    const apps = allApps.filter(appsFilter).sort(sortByStateAndName);

    return (
        <>
            <AppFilterBar />
            <WithScrollbarContainer hasFilter>
                {apps.map(app => (
                    <App key={`${app.name}-${app.source}`} app={app} />
                ))}
            </WithScrollbarContainer>

            <ConfirmLaunchDialog />
            <InstallOtherVersionDialog />
            <ReleaseNotesDialog />
        </>
    );
};
