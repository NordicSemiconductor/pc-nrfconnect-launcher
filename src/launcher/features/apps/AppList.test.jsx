/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import { LOCAL, OFFICIAL } from '../../../ipc/sources';
import render, { createPreparedStore } from '../../../testrenderer';
import { setAllShownSources } from '../filter/filterSlice';
import AppList from './AppList';
import {
    checkEngineAndLaunch,
    installDownloadableApp,
    removeDownloadableApp,
} from './appsEffects';
import {
    installDownloadableAppStarted,
    loadDownloadableAppsSuccess,
    removeDownloadableAppStarted,
    upgradeDownloadableAppStarted,
} from './appsSlice';

// Do not render react-bootstrap components in tests
jest.mock('react-bootstrap', () => ({
    Modal: 'Modal',
    Button: 'Button',
    ModalHeader: 'ModalHeader',
    ModalFooter: 'ModalFooter',
    ModalBody: 'ModalBody',
    ModalTitle: 'ModalTitle',
}));

jest.mock('../../features/filter/AppFilterBar', () => 'div');
jest.mock('../../features/releaseNotes/ReleaseNotesDialog', () => 'div');
jest.mock('../../util/mainConfig', () => () => ({ version: '6.1.0' }));

jest.mock('../../features/apps/appsEffects');

const localApp = {
    name: 'local',
    source: LOCAL,
    displayName: 'Local App',
    description: 'An local app',
    isInstalled: true,
    isDownloadable: false,

    currentVersion: '2.0.0',

    engineVersion: '6.1.0',
    path: '',
    iconPath: '',
    shortcutIconPath: '',
};

const updateableApp = {
    name: 'appA',
    source: OFFICIAL,
    displayName: 'App A',
    description: 'appA description',
    isInstalled: true,
    isDownloadable: true,
    currentVersion: '1.2.3',
    latestVersion: '1.2.4',
    upgradeAvailable: true,

    engineVersion: '6.1.0',
    url: '',
    path: '',
    iconPath: '',
    shortcutIconPath: '',
};

const uninstalledApp = {
    name: 'appB',
    source: OFFICIAL,
    displayName: 'App B',
    description: 'appB description',
    isInstalled: false,
    isDownloadable: true,

    engineVersion: '6.1.0',
    url: '',
    latestVersion: '',
};

const installedApp = {
    name: 'appC',
    source: OFFICIAL,
    displayName: 'App C',
    description: 'appC description',
    isInstalled: true,
    isDownloadable: true,
    currentVersion: '1.2.3',
    latestVersion: '1.2.3',

    engineVersion: '6.1.0',
    url: '',
    path: '',
    iconPath: '',
    shortcutIconPath: '',
};

describe('AppList', () => {
    it('should render without any apps', () => {
        expect(render(<AppList />).baseElement).toMatchSnapshot();
    });

    it('should render local, not-installed, installed, and upgradable apps', () => {
        expect(
            render(<AppList />, [
                setAllShownSources(new Set([OFFICIAL, LOCAL])),
                loadDownloadableAppsSuccess({
                    downloadableApps: [
                        uninstalledApp,
                        installedApp,
                        updateableApp,
                        localApp,
                    ],
                }),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should disable buttons and display "Installing..." button text while installing an app', () => {
        expect(
            render(<AppList />, [
                setAllShownSources(new Set([OFFICIAL, LOCAL])),
                loadDownloadableAppsSuccess({
                    downloadableApps: [uninstalledApp],
                }),
                installDownloadableAppStarted(uninstalledApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should disable buttons while removing an app', () => {
        expect(
            render(<AppList />, [
                setAllShownSources(new Set([OFFICIAL, LOCAL])),
                loadDownloadableAppsSuccess({
                    downloadableApps: [installedApp],
                }),
                removeDownloadableAppStarted(installedApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    fit('should disable buttons and display "Updating..." while updating an app', () => {
        expect(
            render(<AppList />, [
                setAllShownSources(new Set([OFFICIAL, LOCAL])),
                loadDownloadableAppsSuccess({
                    downloadableApps: [updateableApp],
                }),
                upgradeDownloadableAppStarted(updateableApp),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('should invoke installDownloadableApp with app name and source when install button is clicked', () => {
        installDownloadableApp.mockReturnValue({ type: 'an action' });

        const wrapper = mount(
            <Provider
                store={createPreparedStore([
                    setAllShownSources(new Set([OFFICIAL, LOCAL])),
                    loadDownloadableAppsSuccess({
                        downloadableApps: [uninstalledApp],
                    }),
                ])}
            >
                <AppList />
            </Provider>
        );
        wrapper
            .find(`[title="Install ${uninstalledApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(installDownloadableApp).toHaveBeenCalledWith({
            name: uninstalledApp.name,
            source: uninstalledApp.source,
        });
    });

    it('should invoke removeDownloadableApp with app name and source when remove button is clicked', () => {
        removeDownloadableApp.mockReturnValue({ type: 'an action' });

        const wrapper = mount(
            <Provider
                store={createPreparedStore([
                    setAllShownSources(new Set([OFFICIAL, LOCAL])),
                    loadDownloadableAppsSuccess({
                        downloadableApps: [installedApp],
                    }),
                ])}
            >
                <AppList />
            </Provider>
        );

        wrapper.find(`button.dropdown-toggle`).first().simulate('click');
        wrapper
            .find(`[title="Remove ${installedApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(removeDownloadableApp).toHaveBeenCalledWith({
            name: installedApp.name,
            source: installedApp.source,
        });
    });

    it('should invoke checkEngineAndLaunch with given app item when Open is clicked', () => {
        checkEngineAndLaunch.mockReturnValue({ type: 'an action' });

        const wrapper = mount(
            <Provider
                store={createPreparedStore([
                    setAllShownSources(new Set([OFFICIAL, LOCAL])),
                    loadDownloadableAppsSuccess({
                        downloadableApps: [installedApp],
                    }),
                ])}
            >
                <AppList />
            </Provider>
        );

        wrapper
            .find(`[title="Open ${installedApp.displayName}"]`)
            .first()
            .simulate('click');

        expect(checkEngineAndLaunch).toHaveBeenCalledWith(
            expect.objectContaining(installedApp)
        );
    });
});
