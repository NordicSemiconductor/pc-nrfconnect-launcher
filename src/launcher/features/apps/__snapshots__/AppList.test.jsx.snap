// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppList should disable buttons and display "Updating..." while updating an app 1`] = `
<body>
  <div>
    <div />
    <div
      class="with-scrollbar filter-adjusted-height"
    >
      <div
        class="content-container"
      >
        <div
          class="list-group-item"
        >
          <div
            class="py-1 row no-gutters"
          >
            <div
              class="d-flex align-items-start my-2 mr-3 col-auto"
            >
              <div
                class="core-app-icon"
              >
                <div
                  class="icon-replacement"
                />
              </div>
            </div>
            <div
              class="col"
            >
              <div
                class="h8"
              >
                App A
              </div>
              <div
                class="small text-muted"
              >
                appA description
              </div>
              <div
                class="small text-muted-more"
              >
                official
                , v
                1.2.3
                 (v
                1.2.4
                 available)
              </div>
            </div>
            <div
              class="d-flex align-items-center my-3 pl-3 ml-auto col-auto"
            >
              <div
                class="wide-btns btn-toolbar"
                role="toolbar"
              >
                <button
                  class="btn btn-outline-primary"
                  disabled=""
                  title="Update App A"
                  type="button"
                >
                  Updating...
                </button>
                <button
                  class="btn btn-primary"
                  disabled=""
                  title="Open App A"
                  type="button"
                >
                  Open
                </button>
                <div
                  class="dropdown"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-outline-primary"
                    type="button"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</body>
`;
